pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,364 @@
+package statefetcher
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+
+	"github.com/nspcc-dev/neo-go/pkg/config"
+	"github.com/nspcc-dev/neo-go/pkg/core/statesync"
+	gio "github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/services/helpers/neofs"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	"github.com/nspcc-dev/neofs-sdk-go/container"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+// Ledger is the interface for statefetcher.
+type Ledger interface {
+	GetConfig() config.Blockchain
+	HeaderHeight() uint32
+	AddContractStorageData(kv []statesync.KeyValue, syncHeight uint32, expectedRoot util.Uint256) error
+	GetLastStoredKey() []byte
+}
+
+type kvPair struct {
+	statesync.KeyValue
+	err error
+}
+
+// Service fetches contract storage state from NeoFS.
+type Service struct {
+	basic             neofs.BasicService
+	isActive          atomic.Bool
+	isShutdown        atomic.Bool
+	cfg               config.NeoFSStateFetcher
+	stateSyncInterval uint32
+	lastStateIndex    uint32
+	lastStateOid      oid.ID
+
+	chain Ledger
+	log   *zap.Logger
+	wg    sync.WaitGroup
+
+	shutdownCallback func()
+}
+
+// New creates a new Service instance.
+func New(chain Ledger, cfg config.NeoFSStateFetcher, stateSyncInterval int, logger *zap.Logger, shutdownCallback func()) (*Service, error) {
+	if !cfg.Enabled {
+		return &Service{}, nil
+	}
+	if cfg.Timeout <= 0 {
+		cfg.Timeout = neofs.DefaultTimeout
+	}
+	if cfg.StateAttribute == "" {
+		cfg.StateAttribute = neofs.DefaultStateAttribute
+	}
+	if cfg.KeyValueBatchSize <= 0 {
+		cfg.KeyValueBatchSize = neofs.DefaultKvBatchSize
+	}
+
+	basic, err := neofs.New(cfg.NeoFSService)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %w", err)
+	}
+	sfs := Service{
+		basic:             basic,
+		log:               logger,
+		cfg:               cfg,
+		chain:             chain,
+		shutdownCallback:  shutdownCallback,
+		stateSyncInterval: uint32(stateSyncInterval),
+	}
+	if sfs.stateSyncInterval <= 0 {
+		sfs.stateSyncInterval = config.DefaultStateSyncInterval
+	}
+	var (
+		containerID  cid.ID
+		containerObj container.Container
+	)
+	sfs.basic.Ctx, sfs.basic.CtxCancel = context.WithCancel(context.Background())
+	if err = sfs.basic.Pool.Dial(context.Background()); err != nil {
```

Comment:
Can we remove all `basic` mentions from these (and similar) cases? BasicService is inlined field, so `basic` should not be needed.